### PR TITLE
Sensors in the atmosphere

### PR DESCRIPTION
**Pull request by schunkes**
# Description

This PR adds functionality which lets Experiment classes that support atmospheres add an external medium to the sensors, which are placed inside the atmosphere's volume.

# Checklist

- [x] The code follows the relevant coding guidelines
- [x] The code generates no new warnings
- [x] The code is appropriately documented
- [x] The code is tested to prove its function
- [x] The feature branch is rebased on the current state of the `main` branch
- [x] I give permission that the Eradiate project may redistribute my contributions under the terms of its license

### CODE REVIEW

File: src/eradiate/experiments/_rami4atm.py [hunk 1]
```diff
@@ -43,6 +44,12 @@ class Rami4ATMExperiment(EarthObservationExperiment):
     * if neither atmosphere nor canopy are defined, the target is set to
       [0, 0, 0].
 
+    If a sensor is placed inside the atmosphere's volume, the experiment will attempt to specify
+    the surrounding medium of that sensor. If a :class:`MultiRadiancemeterMeasure` with subsensors
+    which are partially inside and outside of the atmosphere is created, the experiment will raise
+    a ValueError. In this case please use two separate measures and group the subsensors which are
+    inside and outside of the atmosphere.
+
```
Comment (leroyvn) [inline src/eradiate/experiments/_rami4atm.py:52]: Same comment as for the 1D module.

**Review by leroyvn — CHANGES_REQUESTED** [review]
Hi @schunkes, I made a bunch of comments and suggestions for a cleaner implementation. I see overall two important issues which I tried to address:

1. The code is lengthy and hard to understand, but a bounding box abstraction will make it more compact, testable and legible.
2. This implementation relies on an instance method which, by nature, is not meant for reuse unless you subclass (not doable in our case). My suggestions address this by making the duplicate component a pure function.

File: src/eradiate/experiments/_onedim.py [hunk 2]
```diff
@@ -93,14 +99,79 @@ def kernel_dict(self, ctx: KernelDictContext) -> KernelDict:
...
+                self.surface,
+                self.illumination,
+                *self.measures,
+                self.integrator,
+                ctx=ctx
+            )
 
         return result
 
+    def _is_measure_inside_atm(self, measure, ctx):
```
[the remark this answers is not in the document]
Comment (schunkes) [reply src/eradiate/experiments/_onedim.py]: I did add a method to the Atmosphere class. I agree that this kind of method makes sense and could be useful in the future as well.

File: src/eradiate/contexts.py [hunk 3]
```diff
@@ -305,6 +305,15 @@ class KernelDictContext(Context):
         default="True",
     )
 
+    atmosphere_medium_id: t.Optional[str] = documented(
+        attr.ib(default=None, converter=attr.converters.optional(str)),
+        doc="If relevant, value which must be used as the surrounding medium for sensors, which"
+        "are positioned inside the atmosphere volume.",
```
Comment (leroyvn) [inline src/eradiate/contexts.py:311]: Let's also broaden the scope of this doc:
```suggestion
        doc="If relevant, a kernel medium ID to reference by other kernel components.",
```

File: src/eradiate/experiments/_onedim.py [hunk 4]
```diff
@@ -10,12 +10,38 @@
...
 from ..scenes.measure._core import MeasureFlags
 from ..scenes.surface import LambertianSurface, Surface, surface_factory
 from ..units import unit_context_config as ucc
 
 
+def measure_inside_atmosphere(atmosphere, measure, ctx):
+    """
+    Evaluate whether a sensor is placed within an atmosphere.
+
+    Raises a ValueError if calld with a :class:`MultiRadiancemeterMeasure` with
```
Comment (leroyvn) [inline src/eradiate/experiments/_onedim.py:23]:
```suggestion
    Raises a ValueError if called with a :class:`.MultiRadiancemeterMeasure` with
```

File: src/eradiate/experiments/_onedim.py [hunk 5]
```diff
@@ -10,12 +10,38 @@
...
+    origins that are inside as well as outside of the atmosphere.
```
Comment (leroyvn) [inline src/eradiate/experiments/_onedim.py:24]:
```suggestion
    origins both inside and outside of the atmosphere.
```

File: src/eradiate/experiments/_onedim.py [hunk 6]
```diff
@@ -10,12 +10,38 @@
...
+    """
+    bbox = atmosphere.eval_bbox(ctx)
+
+    if isinstance(measure, MultiRadiancemeterMeasure):
+        if all(bbox.contains(origin) for origin in measure.origins):
+            return True
+        elif not any(bbox.contains(origin) for origin in measure.origins):
+            return False
```
Comment (leroyvn) [inline src/eradiate/experiments/_onedim.py:32]: Still with the idea of not evaluating twice:
```suggestion
        bbox_contains = [bbox.contains(origin) for origin in measure.origins]
        if all(bbox_contains):
            return True
        elif not any(bbox_contains):
            return False
```

File: src/eradiate/experiments/_onedim.py [hunk 7]
```diff
@@ -10,12 +10,38 @@
...
+        else:
+            raise ValueError(
+                "Inconsistent placement of Multiradiancemeter origins. "
+                "Either all sensors must be placed inside the participating "
+                "medium or all must be placed outside."
```
Comment (leroyvn) [inline src/eradiate/experiments/_onedim.py:37]: How about a briefer
```suggestion
                "Inconsistent placement of MultiRadiancemeterMeasure origins. "
                "All origins must lie either inside or outside of the atmosphere."
```

**Review by leroyvn — COMMENTED** [review]
Hi @schunkes, this looks good to me—I just had a couple of comments, small details. When this is taken care of, we can squash + merge!

**Comment by leroyvn** [comment]
I made a few more tiny updates, let's merge. Thanks @schunkes!

**Pull request merged**